src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,48 +3,193 @@
+        if (index < size / 2) {
+            newNode = first;
+            for (int i = 0; i < index; i++) {
+                newNode = newNode.next;
+            }
+            return newNode;
+        } else {
+            newNode = last;
+            for (int i = size - 1; i > index; i--) {
+                newNode = newNode.prev;
+            }
+            return newNode;
+        }
+    }
+
+    public void addFirst(T value) {
+        newNode = new Node(value);
+        first = last = newNode;
+    }
+
+    public void addFront(T value) {
+        newNode = new Node(value);
+        first.prev = newNode;
+        newNode.next = first;
+        newNode.prev = null;
+        first = newNode;
+    }
+
+    public void addBack(T value) {
+        newNode = new Node(value);
+        newNode.prev = last;
+        last.next = newNode;
+        newNode.next = null;
+        last = newNode;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == 0) {
+            addFirst(value);
+        } else {
+            addBack(value);
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (size < index || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (index == 0) {
+            if (size == 0) {
+                addFirst(value);
+            } else {
+                addFront(value);
+            }
+        }
+        if (index > 0) {
+            if (index == size) {
+                addBack(value);
+            }
+            newNode = new Node(value);
+            nodeByIndex(index - 1).next = newNode;
+            newNode.next = nodeByIndex(index);
+            newNode.prev = nodeByIndex(index - 1);
+            nodeByIndex(index).prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (size <= index || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        return nodeByIndex(index).item;
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (size <= index || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        nodeByIndex(index).item = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> nodeToBeRemoved = null;
+        if (size <= index || index < 0) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (size == 1) {
+            first = last = null;
+            size--;
+            return null;
+        }
+        if (index == 0) {
+            nodeToBeRemoved = first;
+            first.next.prev = null;
+            first = null;
+            first = nodeToBeRemoved.next;
+        }
+        if (index > 0) {
```

Comment:
d owe need this `if`?

Reply:
looks like we don't need... but it is not works if I remove it((...

Reply:
I removed it in add(value, index), but can't remove it in addAll()

Reply:
we can white else {...

Reply:
done